runner/organizations_test.go

```diff
@@ -22,170 +22,747 @@ import (
+	s.Fixtures.PoolMgrMock.On("Start").Return(nil)
+	s.Fixtures.PoolMgrCtrlMock.On("CreateOrgPoolManager", s.Fixtures.AdminContext, s.Fixtures.Org, s.Fixtures.Providers, s.Fixtures.StoreMock).Return(s.Fixtures.PoolMgrMock, nil)
 
-	org, err := runner.CreateOrganization(adminCtx, createOrgParams)
-	require.NotNil(t, err)
-	require.Equal(t, params.Organization{}, org)
-	require.Regexp(t, "validating params: missing repo name", err.Error())
+	// call tested function
+	org, err := s.Runner.CreateOrganization(s.Fixtures.AdminContext, s.Fixtures.CreateOrgParams)
+
+	// assertions
+	s.Fixtures.StoreMock.AssertExpectations(s.T())
+	s.Fixtures.PoolMgrMock.AssertExpectations(s.T())
+	s.Fixtures.PoolMgrCtrlMock.AssertExpectations(s.T())
+	require.Equal(s.T(), s.Fixtures.Org, org)
+	require.Nil(s.T(), err)
 }
 
-func TestCreateOrganizationMissingCredentials(t *testing.T) {
-	adminCtx := auth.GetAdminContext()
-	createOrgParams := params.CreateOrgParams{
-		Name:            "test",
-		CredentialsName: "test",
-		WebhookSecret:   "test",
+func (s *OrgTestSuite) TestCreateOrganizationErrUnauthorized() {
+	org, err := s.Runner.CreateOrganization(s.Fixtures.Context, s.Fixtures.CreateOrgParams)
+
+	require.Equal(s.T(), params.Organization{}, org)
+	require.Equal(s.T(), runnerErrors.ErrUnauthorized, err)
+}
+
+func (s *OrgTestSuite) TestCreateOrganizationInvalidParams() {
+	// setup environment specific for this test
+	s.Fixtures.CreateOrgParams = params.CreateOrgParams{}
+
+	// call tested function
+	org, err := s.Runner.CreateOrganization(s.Fixtures.AdminContext, s.Fixtures.CreateOrgParams)
+
+	// assertions
+	require.Equal(s.T(), params.Organization{}, org)
+	require.Regexp(s.T(), "validating params: missing org name", err.Error())
+}
+
+func (s *OrgTestSuite) TestCreateOrganizationMissingCredentials() {
+	s.Fixtures.CreateOrgParams.CredentialsName = "not-existent-creds-name"
+
+	org, err := s.Runner.CreateOrganization(s.Fixtures.AdminContext, s.Fixtures.CreateOrgParams)
+
+	require.Equal(s.T(), params.Organization{}, org)
+	require.Equal(s.T(), runnerErrors.NewBadRequestError("credentials %s not defined", s.Fixtures.CreateOrgParams.CredentialsName), err)
+}
+
+func (s *OrgTestSuite) TestCreateOrganizationOrgFetchFailed() {
+	s.Fixtures.StoreMock.On("GetOrganization", s.Fixtures.AdminContext, s.Fixtures.CreateOrgParams.Name).Return(s.Fixtures.Org, s.Fixtures.ErrMock)
+
+	org, err := s.Runner.CreateOrganization(s.Fixtures.AdminContext, s.Fixtures.CreateOrgParams)
+
+	s.Fixtures.StoreMock.AssertExpectations(s.T())
+	require.Equal(s.T(), params.Organization{}, org)
+	require.Equal(s.T(), fmt.Sprintf("fetching org: %s", s.Fixtures.ErrMock.Error()), err.Error())
+}
+
+func (s *OrgTestSuite) TestCreateOrganizationAlreadyExists() {
+	s.Fixtures.CreateOrgParams.Name = "not-existent-name"
```

Comment:
Why change the name here, if we'll be returning the fixture anyway? We're setting up the mock to call ```GetOrganization``` with ```not-existent-name``` and returning an org from the fixture, which does not include a name. If you skip this, you should get the same result.

I am wondering if we should have a proper testing implementation of the database ```common.Store{}``` interface that actually creates those entities in memory, as normal store would, and return proper objects with all fields set. Or maybe just use the ```sql``` implementation with the ```sqlite``` backend, given the self contained nature of ```sqlite```.

The issue with mocking everything (duck typing) is that you run the risk of returning expectations that are not correct. Mocks may be outdated, or expectations may change. You may miss errors. It requires a lot of discipline to ensure mocks are correct or we risk tests lying to us. I am not suggesting that we never use mocks, but we should do so sparingly. Only use them where we have to, like when we need to call into the GitHub API or into the pool manager interface (which is somewhat opaque to this package).